src/backend/distributed/commands/schema.c

```diff
@@ -53,76 +85,48 @@ PreprocessDropSchemaStmt(Node *node, const char *queryString,
 						 ProcessUtilityContext processUtilityContext)
 {
 	DropStmt *dropStatement = castNode(DropStmt, node);
-	Relation pgClass = NULL;
-	HeapTuple heapTuple = NULL;
-	SysScanDesc scanDescriptor = NULL;
-	ScanKeyData scanKey[1];
-	int scanKeyCount = 1;
-	Oid scanIndexId = InvalidOid;
-	bool useIndex = false;
+	Assert(dropStatement->removeType == OBJECT_SCHEMA);
 
-	if (dropStatement->behavior != DROP_CASCADE)
+	if (!ShouldPropagate())
 	{
 		return NIL;
 	}
 
-	Value *schemaValue = NULL;
-	foreach_ptr(schemaValue, dropStatement->objects)
+	EnsureCoordinator();
+
+	List *distributedSchemas = FilterDistributedSchemas(dropStatement->objects);
+
+	if (list_length(distributedSchemas) < 1)
 	{
-		const char *schemaString = strVal(schemaValue);
-		Oid namespaceOid = get_namespace_oid(schemaString, true);
+		return NIL;
+	}
 
-		if (namespaceOid == InvalidOid)
-		{
-			continue;
-		}
+	EnsureSequentialModeForSchemaDDL();
 
-		pgClass = table_open(RelationRelationId, AccessShareLock);
+	Value *schemaVal = NULL;
+	foreach_ptr(schemaVal, distributedSchemas)
+	{
+		InvalidateFKeysForDistributedTablesInSchema(strVal(schemaVal));
```

Comment:
Looking at the code this replaces on the original line 115 we had an early escape once the graph has been invalidated. With multiple schema's do we need to invalidate the graph twice?

Has this been considered as an option?

I don't think it is a huge performance problem, given this is non-frequent executed code. If the readability decreases we can keep it like this, but a comment that this behaviour is expected would be nice.

Alternatively this could be refactored slightly more in a function that simply checks if the schema has any tables referencing something here. When it does an early escape with a bool will suffice.

Once we have any schema that has references that require invalidation we can simply call `MarkInvalidateForeignKeyGraph` from here.

This all is under the assumption that the original code was _only_ having the side effect of invalidating the fkey graph.

As said earlier, this is by no means required, but reasoning why we don't do this will help our future selfs.